File: context/data/apache_arrow/examples/pipeline_parallelism.py
import queue
import threading
import time
import pyarrow.flight as fl

# Configuration to match flight_server.py
LOCATION = "grpc://localhost:8815"
# The server expects this path in the descriptor to find the table
DATA_PATH = "large_table"
QUEUE_SIZE = 4

# Queues for the pipeline stages
q_network_to_compute = queue.Queue(maxsize=QUEUE_SIZE)
q_compute_to_sink = queue.Queue(maxsize=QUEUE_SIZE)


def stage_a_network():
    """Stage A: Fetch batches from the network (I/O Bound)"""
    print("[Stage A] Connecting to Flight server...")
    try:
        with fl.FlightClient(LOCATION) as client:
            # 1. Get FlightInfo to retrieve the Ticket
            descriptor = fl.FlightDescriptor.for_path(DATA_PATH)
            info = client.get_flight_info(descriptor)

            # 2. Use the ticket from the first endpoint
            ticket = info.endpoints[0].ticket
            reader = client.do_get(ticket)

            for i, batch_chunk in enumerate(reader):
                batch = batch_chunk.data
                print(f"[Stage A] Received batch #{i + 1} ({batch.num_rows} rows)")
                q_network_to_compute.put((i + 1, batch))

        print("[Stage A] All batches received.")
    except Exception as e:
        print(f"[Stage A] Error: {e}")
    finally:
        q_network_to_compute.put(None)


def stage_b_compute():
    """Stage B: Process data (CPU Bound)"""
    while True:
        item = q_network_to_compute.get()
        if item is None:
            print("[Stage B] No more data. Stopping.")
            q_compute_to_sink.put(None)
            break

        batch_id, batch = item
        print(f"  [Stage B] Processing batch #{batch_id}...")

        # Simulate heavy processing
        time.sleep(0.6)

        # Example transformation: count nulls (like your previous example)
        null_counts = {batch.schema.names[i]: batch.column(i).null_count for i in range(batch.num_columns)}

        q_compute_to_sink.put((batch_id, null_counts))


def stage_c_sink():
    """Stage C: Save results (I/O Bound)"""
    while True:
        item = q_compute_to_sink.get()
        if item is None:
            print("[Stage C] Task complete.")
            break

        batch_id, results = item
        print(f"    [Stage C] Writing results of batch #{batch_id}: {results}")

        # Simulate writing to a database or logs
        time.sleep(0.4)


def run_pipeline():
    print(f"Starting Pipeline Parallelism (Location: {LOCATION})")
    print("-" * 50)

    threads = [
        threading.Thread(target=stage_a_network),
        threading.Thread(target=stage_b_compute),
        threading.Thread(target=stage_c_sink)
    ]

    start_time = time.time()
    for t in threads: t.start()
    for t in threads: t.join()

    print("-" * 50)
    print(f"Total pipeline time: {time.time() - start_time:.2f} seconds.")


if __name__ == "__main__":
    run_pipeline()

File: context/data/apache_arrow/examples/pipeline_parallelism.go
package main

import (
	"context"
	"fmt"
	"io"
	"log"
	"strings"
	"time"

	"github.com/apache/arrow/go/v15/arrow"
	"github.com/apache/arrow/go/v15/arrow/flight"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

const (
	location  = "localhost:8815"
	dataPath  = "large_table"
	queueSize = 4
)

type BatchItem struct {
	ID    int
	Batch arrow.Record
}

type ResultItem struct {
	ID      int
	Results map[string]int
}

func main() {
	fmt.Printf("Starting Go Pipeline Parallelism (Location: %s)\n", location)
	fmt.Println(strings.Repeat("-", 50))

	startTime := time.Now()

	qNetworkToCompute := make(chan BatchItem, queueSize)
	qComputeToSink := make(chan ResultItem, queueSize)

	// Stage A: Network
	go func() {
		if err := stageANetwork(qNetworkToCompute); err != nil {
			log.Printf("[Stage A] Error: %v", err)
		}
		close(qNetworkToCompute)
	}()

	// Stage B: Compute
	go func() {
		stageBCompute(qNetworkToCompute, qComputeToSink)
		close(qComputeToSink)
	}()

	// Stage C: Sink (Blocking in main)
	stageCSink(qComputeToSink)

	fmt.Println(strings.Repeat("-", 50))
	fmt.Printf("Total pipeline time: %.2f seconds.\n", time.Since(startTime).Seconds())
}

func stageANetwork(out chan<- BatchItem) error {
	fmt.Println("[Stage A] Connecting to Flight server...")

	conn, err := grpc.Dial(location, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return err
	}
	defer conn.Close()

	client := flight.NewFlightClient(conn)

	// 1. Get FlightInfo to retrieve the Ticket
	fd := &flight.FlightDescriptor{
		Type: flight.DescriptorPath,
		Path: []string{dataPath},
	}
	info, err := client.GetFlightInfo(context.Background(), fd)
	if err != nil {
		return err
	}

	// 2. Use the ticket from the first endpoint
	ticket := info.Endpoint[0].Ticket
	stream, err := client.DoGet(context.Background(), ticket)
	if err != nil {
		return err
	}

	reader, err := flight.NewRecordReader(stream)
	if err != nil {
		return err
	}
	defer reader.Release()

	batchID := 1
	for reader.Next() {
		batch := reader.Record()
		batch.Retain() // Retain because it's sent over channel
		fmt.Printf("[Stage A] Received batch #%d (%d rows)\n", batchID, batch.NumRows())
		out <- BatchItem{ID: batchID, Batch: batch}
		batchID++
	}

	if reader.Err() != nil && reader.Err() != io.EOF {
		return reader.Err()
	}

	fmt.Println("[Stage A] All batches received.")
	return nil
}

func stageBCompute(in <-chan BatchItem, out chan<- ResultItem) {
	for item := range in {
		fmt.Printf("  [Stage B] Processing batch #%d...\n", item.ID)

		// Simulate heavy processing
		time.Sleep(600 * time.Millisecond)

		// Example transformation: count nulls
		nullCounts := make(map[string]int)
		schema := item.Batch.Schema()
		for i := 0; i < int(item.Batch.NumCols()); i++ {
			name := schema.Field(i).Name
			nullCount := item.Batch.Column(i).NullN()
			nullCounts[name] = nullCount
		}

		item.Batch.Release() // Release after processing
		out <- ResultItem{ID: item.ID, Results: nullCounts}
	}
	fmt.Println("[Stage B] No more data. Stopping.")
}

func stageCSink(in <-chan ResultItem) {
	for item := range in {
		fmt.Printf("    [Stage C] Writing results of batch #%d: %v\n", item.ID, item.Results)

		// Simulate writing to a database or logs
		time.Sleep(400 * time.Millisecond)
	}
	fmt.Println("[Stage C] Task complete.")
}
